python/python/lance/dataset.py

```diff
@@ -205,6 +205,153 @@ def retry_timeout(self, timeout: timedelta) -> "MergeInsertBuilder":
+
+        Returns
+        -------
+        str
+            A formatted string representation of the execution plan.
+
+        Examples
+        --------
+        >>> import lance
+        >>> import pyarrow as pa
+        >>> data = pa.table([
+        ...     pa.array([1, 2, 3]),
+        ...     pa.array(['alice', 'bob', 'charlie']),
+        ...     pa.array([100.0, 200.0, 300.0])
+        ... ], names=['id', 'name', 'score'])
+        >>> dataset = lance.write_dataset(data, "memory://test_dataset")
+
+        >>> # Using default dataset schema
+        >>> builder = dataset.merge_insert("id")
+        >>> builder = builder.when_matched_update_all().when_not_matched_insert_all()
+        >>> plan = builder.explain_plan()  # Uses dataset schema
+        >>> print(plan) # doctest: +ELLIPSIS
+        MergeInsert: on=[id], when_matched=UpdateAll, when_not_matched=InsertAll, ...
+          CoalescePartitionsExec
+            ProjectionExec: expr=[_rowaddr@1 as _rowaddr, id@2 as id, ...]
+              ProjectionExec: expr=[id@1 IS NOT NULL as __common_expr_1, ...]
+                CoalesceBatchesExec: target_batch_size=...
+                  HashJoinExec: mode=CollectLeft, join_type=Right, ...
+                    LanceRead: uri=test_dataset/data, projection=[id], ...
+                    RepartitionExec: ...
+                      StreamingTableExec: partition_sizes=1, ...
+        <BLANKLINE>
+
+        >>> # Or with explicit schema
+        >>> source_schema = pa.schema([
+        ...     pa.field("id", pa.int64()),
+        ...     pa.field("name", pa.string()),
+        ...     pa.field("score", pa.float64())
+        ... ])
+        >>> plan = builder.explain_plan(source_schema, verbose=True)
+        >>> print(plan) # doctest: +ELLIPSIS
+        MergeInsert: on=[id], when_matched=UpdateAll, when_not_matched=InsertAll, ...
+          CoalescePartitionsExec
+            ProjectionExec: expr=[_rowaddr@1 as _rowaddr, id@2 as id, ...]
+              ProjectionExec: expr=[id@1 IS NOT NULL as __common_expr_1, ...]
+                CoalesceBatchesExec: target_batch_size=...
+                  HashJoinExec: mode=CollectLeft, join_type=Right, ...
+                    ...
+        """
+        if schema is None:
+            return super(MergeInsertBuilder, self).explain_plan(verbose=verbose)
+        else:
+            return super(MergeInsertBuilder, self).explain_plan(schema, verbose=verbose)
+
+    def analyze_plan(
+        self,
+        data_obj: ReaderLike,
+        *,
+        schema: Optional[pa.Schema] = None,
+    ) -> str:
+        """
+        Generate the execution plan and analyze its performance with real data.
+
+        This method creates the execution plan using the provided source data,
+        executes it to collect performance metrics, and returns the analysis
+        as a formatted string. This is useful for understanding the actual
+        performance characteristics of the merge insert operation with your data.
+
```

Comment:
Should we add a warning that this will execute the operation and insert new data, modifying the dataset?  I feel like it's slightly different from query analyze_plan in that regard.

Reply:
This wasn't very clear, but I actually designed this *not* to modify the dataset. It runs the operation and writes data files. But it doesn't attempt to commit the transaction. (The execution plan itself doesn't do the commit.)

Reply:
Ah, I missed that, and agree it is a good idea.